### ISSUE TITLE
plugin: markdown-plugin/text-plugin support "extensions" option

### ISSUE THREAD

**Issue by azu** [textlint/textlint#531]
markdown-plugin/text-plugin support only hard-coded specific extension.

For example, [@textlint/textlint-plugin-markdown](https://github.com/textlint/textlint/tree/102d5683ea7172ead702acf00e9fc111a53637d3/packages/%40textlint/textlint-plugin-markdown) support `[".md", ".markdown", ".mdown", ".mkdn", ".mkd", ".mdwn", ".mkdown", ".ron"]` extensions.

Some user want to treat other file extension as markdown.

- https://twitter.com/garin54/status/1011904067470176256
   - Want to treat `.hown` as Markdown

We should extend this supported file extendsion by option.

**Comment by azu** [comment]
## Proposal A

## Tasks

- [ ] Add `extensions` option to [@textlint/textlint-plugin-markdown](https://github.com/textlint/textlint/tree/102d5683ea7172ead702acf00e9fc111a53637d3/packages/%40textlint/textlint-plugin-markdown) and [@textlint/textlint-plugin-text](https://github.com/textlint/textlint/tree/master/packages/%40textlint/textlint-plugin-text)
  - Merge the `extensions` to defaut `availableExtensions()`.
The user can add supported extentions by following:

```json5
{
    "plugins": {
        "@textlint/markdown": {
           "extensions": [".hown", ".custom-file-type"]
        }
    }
}
```

## 📝 Techinically Notes

textlint's plugins accept `options` as constructor arguments.
https://github.com/textlint/textlint/blob/master/packages/textlint/test/pluguins/PluginOption-test.js

## 🔥 Problem 

- `static availableExtensions` is static method.
  - It can not receive `extensions` option.

https://github.com/textlint/textlint/blob/102d5683ea7172ead702acf00e9fc111a53637d3/packages/%40textlint/textlint-plugin-markdown/src/MarkdownProcessor.js#L11

Context:

textlint collect all `static availableExtensions` before linting.
Becasuse, It would less finding cost.

https://github.com/textlint/textlint/blob/a144570246d0ec89cbe370f80b5e0706e5c0647a/packages/textlint/src/config/plugin-loader.ts
https://github.com/textlint/textlint/blob/442c297c60e77de55f07fea7c75d71ac05a6a03a/packages/textlint/src/engine/textlint-engine-core.ts#L223

**Comment by azu** [comment]
## Proposal B

The user can defined mapping of extension and plugin

`.textlintrc`
```json5
{
    "extensions": [
       { ".hown": @textlint/markdown }
    ],
    plugins: ["@textlint/markdown"]
}
```

**Comment by azu** [comment]
Another view point:

How to does `@textlint/kernel`receive the custom extensions ?

```js
import { TextlintKernel } from "@textlint/kernel";
const kernel = new TextlintKernel();
const options = {
    filePath: "/path/to/file.md",
    ext: ".md",
    plugins: [
        {
            pluginId: "markdown",
            plugin: require("@textlint/textlint-plugin-markdown"),
            options: {
               extensions: [".hown"]
            }
        }
    ],
    rules: [
        {
            ruleId: "no-todo",
            rule: require("textlint-rule-no-todo")
        }
    ]
};
kernel.lintText("TODO: text", options).then(result => {
    assert.ok(typeof result.filePath === "string");
});
```

**Comment by azu** [comment]
In my opion, I'll take Proposal A.
This tasks has separated tasks.

## Tasks

- Create a function or module that return a map of extension and plugin instance from array of object
   - the return object should be sharead between `textlint` and `@textlint/kernel`

```js
createPluginExtensionMap([
        {
            pluginId: "markdown",
            plugin: require("@textlint/textlint-plugin-markdown"),
            options: {
               extensions: [".hown"]
            }
        }
    ]);
/* Maybe PluginMapper class?
{
  ".hown":  {
            pluginId: "markdown",
            plugin: markdownPluginInstance
   },
  ".md":  {
            pluginId: "markdown",
            plugin: markdownPluginInstance,
   },
   // ....
}
*/
```

- Use the mapper instead of `findPluginWithExt` 

https://github.com/textlint/textlint/blob/abd314abe5fe418051e79f1f8978540a9095e9a2/packages/%40textlint/kernel/src/textlint-kernel.ts#L30-L55

**Comment by azu** [comment]
I've implemented Promise A in #537 
We will release it in [11.0.0](https://github.com/textlint/textlint/milestone/3)

**Comment by azu** [comment]
Availabe this feature in textlint 11.0.0

- https://textlint.github.io/blog/2018/07/22/textlint-11.html
